### ISSUE TITLE
CMake-ify H5Z-ZFP

### ISSUE THREAD

**Issue by markcmiller86** [LLNL/H5Z-ZFP#17]
Title says it all.

We should switch to CMake for H5Z-ZFP.

**Comment by mathomp4** [comment]
@markcmiller86 I think I might try this out. 

You might remember me from #11. Well, I finally think I might have a bit of time to try and get H5Z-ZFP working with netCDF. 

I have some ugly `CMakeLists.txt` at the moment that can make a `.a` and `.so`, but it's not for public yet. One issue so far is that ZFP themselves aren't quite doing CMake correctly, so I can't do `find_package(ZFP)`. But baby steps...

**Comment by mathomp4** [comment]
Well, maybe I'm not quite there yet:
```
$ setenv HDF5_PLUGIN_PATH /discover/nobackup/mathomp4/ZFPTest/src/H5Z-ZFP/build/installed/plugin
$ ls $HDF5_PLUGIN_PATH/
libh5zzfp.so*
$ $ nccopy -F 'T,32013,6,3,0,3539053052,1062232653,0,0' stock-J32-1day-c24.geosgcm_prog.20000415_0000z.nc4 test.nc4
NetCDF: HDF error
Location: file nccopy.c; line 1945
```
😄

**Comment by markcmiller86** [comment]
> @markcmiller86 I think I might try this out.

Great. I would welcome the help.

> You might remember me from #11. Well, I finally think I might have a bit of time to try and get H5Z-ZFP working with netCDF.
> 
> I have some ugly `CMakeLists.txt` at the moment that can make a `.a` and `.so`, but it's not for public yet. One issue so far is that ZFP themselves aren't quite doing CMake correctly, so I can't do `find_package(ZFP)`.

Have you reported that [here](https://github.com/LLNL/zfp/issues)?

**Comment by markcmiller86** [comment]
> $ $ nccopy -F 'T,32013,6,3,0,3539053052,1062232653,0,0' stock-J32-1day-c24.geosgcm_prog.20000415_0000z.nc4 test.nc4

Where are you gettting the arguments for the `-F` option? Are you using [this utility](https://h5z-zfp.readthedocs.io/en/latest/h5repack.html#constructing-an-hdf5-cd-values-array)

**Comment by mathomp4** [comment]
> Have you reported that here?
@markcmiller86 Not yet. I'm waiting for my CMake guru of a boss to return from travel. He can tell me what to fix there probably in like 10 minutes. It'd take me a week of CMake debugging. I figure for now I can just hardcode a link to the library.

> Where are you gettting the arguments for the -F option? Are you using this utility

I didn't even know that exists! I think the numbers I was using was from some sort of example I saw on the web.

**Comment by mathomp4** [comment]
@markcmiller86 Okay. I managed to build `print_h5repack_farg` but apparently not correctly. From your examples:
```
% ./print_h5repack_farg zfpmode=3 acc=0.075

Print cdvals for set of ZFP compression paramaters...
    zfpmode=3        set zfp mode (1=rate,2=prec,3=acc,4=expert)
    rate=4                      set rate for rate mode of filter
    acc=0.075           set accuracy for accuracy mode of filter
    prec=11       set precision for precision mode of zfp filter
    minbits=0          set minbits for expert mode of zfp filter
    maxbits=4171       set maxbits for expert mode of zfp filter
    maxprec=64         set maxprec for expert mode of zfp filter
    minexp=-1074        set minexp for expert mode of zfp filter
    help=0                                     this help message

h5repack -f argument...
    -f UD=32013,4,3,0,858993459,1068708659
```
From my utility:
```
$ ./print_h5repack_farg zfpmode=3 acc=0.075
    
Print cdvals for set of ZFP compression paramaters...
    zfpmode=3  set zfp mode (1=rate,2=prec,3=acc,4=expert,5=rev)
    rate=4                      set rate for rate mode of filter
    acc=0.075           set accuracy for accuracy mode of filter
    prec=0        set precision for precision mode of zfp filter
    minbits=0          set minbits for expert mode of zfp filter
    maxbits=0          set maxbits for expert mode of zfp filter
    maxprec=0          set maxprec for expert mode of zfp filter
    minexp=0            set minexp for expert mode of zfp filter
    help=0                                     this help message

h5repack -f argument...
    -f UD=32013,4,3,0,858993459,1068708659
```
The argument at the end is the same (yay!) but the print out is a lot of 0's on my end.

**Comment by markcmiller86** [comment]
I think the output you are seeing is correct. My documentation is slightly wrong. It is from a version of that tool *before* I defaulted everything to zero there.

**Comment by jwsblokland** [comment]
Hello Mark,

The last few days, I implemented a CMake build configuration for building your HDF5 zfp filter:
 - Currently, it only contains building both static library and the plugin. 
 - Implemented the option to disable the Fortran interface, which may be useful if the HDF5 library itself only consists of the C API.
 - Implemented the generation CMake config files such one can easily add the filter in other CMake Projects.
 
As a side note, your current build system still works and can be used. My plan is to send you a pull request in the coming few days. Today I sent a pull request to LLNL/zfp regarding OpenMP dependency in their current CMake build configuration. If this pull request is not accepted as it currently is, it may effect my current CMake build configuration for your filter.

Best regards,
Jan-Willem

**Comment by markcmiller86** [comment]
I would welcome a CMake build system. One thing...are you taking care to not *require* a too recent version of CMake? I don't think anything here really requires any bleeding-edge CMake features and would like to have something that won't wind up requiring would-be users to have to also install an updated CMake.

**Comment by jwsblokland** [comment]
My current implementation requires CMake 3.9 or later. Hopefully, this is version is ok for you. This is related the pull request I submitted to LLNL/zfp in which I slightly modified their CMake build system such that the possible OpenMP dependency is properly propagated to the outside world in their CMake config files. They had a solution but unfortunately that one was not relocatable. Using CMake 3.9 it was possible to create a proper relocatable solution for this OpenMP dependency.

Why this matter for the filter. In the end an user who wants to build the filter should not be bothered with setting the right correct compiler flags when zfp has been built with OpenMP support. CMake should take care of this, which it does in the current configuration.

**Comment by mathomp4** [comment]
> My current implementation requires CMake 3.9 or later. Hopefully, this is version is ok for you. This is related the pull request I submitted to LLNL/zfp in which I slightly modified their CMake build system such that the possible OpenMP dependency is properly propagated to the outside world in their CMake config files. They had a solution but unfortunately that one was not relocatable. Using CMake 3.9 it was possible to create a proper relocatable solution for this OpenMP dependency.

Huzzah for CMake! Also....daaaaaaang CMake 3.9 (from 2017)?! You get all the applause. I am struggling not to force CMake 3.20 on my users because, well, I like being on the bleeding edge. (3.20.0 was released in April, that's sooo olllldddddd! 😃 )

**Comment by jwsblokland** [comment]
Mark and Matthew, 

A moment ago, I sent in my Pull Request about CMake build system.

**Comment by markcmiller86** [comment]
Resolved in #52
